src/storage/lock_manager/deadlock.rs

```diff
@@ -23,83 +23,109 @@ use std::fmt::{self, Display, Formatter};
 use std::rc::Rc;
 use std::sync::Arc;
 use std::time::{Duration, Instant};
+use tikv_util::time;
 use tokio_core::reactor::Handle;
 use tokio_timer::Interval;
 
 // 2 mins
 const TXN_DETECT_INFO_TTL: u64 = 120000;
 
-#[derive(Default)]
+/// Used to detect the deadlock of wait-for-lock in the cluster.
 struct DetectTable {
-    // txn_ts => (lock_ts, Vec<lock_hash>)
-    wait_for_map: HashMap<u64, HashMap<u64, Vec<u64>>>,
+    /// Keeps the DAG of wait-for-lock. Every edge from `txn_ts` to `lock_ts` has a survival time -- `ttl`.
+    /// When checking the deadlock, if the ttl has elpased, the corresponding edge will be removed.
+    /// `last_detect_time` is the start time of the edge. `Detect` requests will refresh it.
+    // txn_ts => (lock_ts => (Vec<lock_hash>, last_detect_time))
+    wait_for_map: HashMap<u64, HashMap<u64, (Vec<u64>, time::Instant)>>,
+
+    now: time::Instant,
```

Comment:
Could we use `max_ts` in the struct `Inner` instead? I do not think using the time based on the local machine's time is a good idea.

Reply:
I will remove the `max_ts` and original expire-timer.😂 Local machine's time is not a problem.

Reply:
If the local clock skews,  the entry may be removed early or late. But it is not a big deal. Because  txns will retry after timeout.
BTW, how do we implement lease read? Is not it use the local clock?